designs/plugins-phase-2.md

```diff
@@ -0,0 +1,105 @@
+
+Boilerplate is not currently in the plugin Universe because it’s language specific and depending on the language, it could change. Some alternatives to this are using a boilerplate path and the plugin could check for its existence, or it could exist in the universe not as a special file but as a list of strings without comments or markers which could be read line by line, thereby adding it to any file. Along with the above `plugin.Universe`, all the raw flags that would be passed to the subcommand need to be passed along across stdin/stdout. And the idea here is to make the plugin generic such that the args are passed as they are without specifying the resource explicitly, which eliminates the need to write a plugin struct to support every possible subcommand like init, create api, and create webhook. 
+
+
+## Configuration
+
+### Project file plugin type & enhanced `layout`
+
+Currently, the `PROJECT` file specifies what base plugin generated the project under a `layout` key (which is the plugin semantic version) in the format: `Plugin.Name() + "/" + Plugin.Version()`.
+
+To support phase 2 plugins, we propose that we modify the "layout" PROJECT file key to either of these options as order matters for certain plugins: 
+    * Ordered list of plugin {name}/{version} or
+    * Ordered map of {name}/{version} to plugin-specific config
+
+Example `PROJECT` file:
+
+```yaml
+version: "3-alpha"
+layout: ["go/v1.0.0", "go.kubebuilder.io/v3-alpha"]
+domain: testproject.org
+repo: github.com/test-inc/testproject
+resources:
+- group: crew
+  kind: Captain
+  version: v1
+plugins:
+  - go.sdk.operatorframework.io/v2-alpha: {
+      type: JSON
+  }
+``` 
+
+## CLI
+
+Idea is to add support for a global --plugins flag that takes an ordered list of plugin names. The subcommand invoked with --plugins executes "downstream" plugins from that list that match those the kubebuilder binary knows about.
+
+Given that we want to pass some initial universe between plugins, that universe should be initialized by the plugin invoked via CLI (the "base" plugin). For example, kubebuilder init --plugins addon will invoke an Init plugin, which passes its generated state to the addon plugin, modifying scaffolded files then writing them to disk.
+
+## Prototype implementation
+
+Kubebuilder feature branch: https://github.com/kubernetes-sigs/kubebuilder/tree/feature/plugins-part-2-electric-boogaloo
+
+## Related issues and PRs
+
+* [Feature request: Plugins Phase 2](https://github.com/kubernetes-sigs/kubebuilder/issues/1378)
+
+## Comments & Questions
+
+* How to handle help text? Are plugins allowed to define their own flags?
```

Comment:
Could  you please add an example over who the consumer will consume and customize the plugins in phase 2? Or are we not covering it in this phase? For example, I am as a consumer(sdk) of kubebuilder plugins, would like to extend the init subcommand/plugin and define that the main.go will be a scaffold in a new path, so that for my tool can do the same scaffolds but in another path. 

Also, would be nice we have here what are the goals and no goals of this EP. So, if it is not a goal of this proposal we would easily know.